### ISSUE TITLE
Improve authentication fetching in tiller-proxy

### ISSUE THREAD

**Issue by migmartri** [kubeapps/kubeapps#895]
See https://github.com/kubeapps/kubeapps/pull/878#discussion_r243023119 for more information.

**Comment by andresmgot** [comment]
Moving the conversation here @migmartri @prydonius

**Comment by andresmgot** [comment]
migmartri 14 hours ago  Member
Does this mean that for every single request made to get the chart we are going to initialize the client and hence read a secret for Kubernetes? If that's the case it seems an unacceptable overload
   @migmartri
migmartri 14 hours ago  Member
We might want to have the ca's preloaded or add an additional cache layer so it is fetched only once?
   @migmartri
migmartri 14 hours ago  Member
Actually just noticed that we are already reading the secret in GetChart which means that the secret wiill be fetched twice in the case of both header and CA are set no?

Would it be possible to preload this or cache it?
   @prydonius
prydonius 14 hours ago  Member
We shouldn't assume in tiller-proxy that the auth header and custom CA Secrets are the same, given they both each have their own secretKeyRefs and if we change the way we handle the Secrets in the future we don't need to worry about this.

I think a cache layer (that caches by key to avoid above issue) that fetches if it isn't loaded might be a good idea - but what happens if the Secret gets updated? When there's a cache hit, we could return the cached value and then spin off a goroutine to get the updated Secret, but that would still cause it to fail the first time it tries to install a chart with an updated auth header/custom CA.
   @migmartri
migmartri 14 hours ago  Member
We shouldn't assume in tiller-proxy that the auth header and custom CA Secrets are the same, given they both each have their own secretKeyRefs and if we change the way we handle the Secrets in the future we don't need to worry about this.
That's fine as long as there is some caching involve.

I think a cache layer (that caches by key to avoid above issue) that fetches if it isn't loaded might be a good idea - but what happens if the Secret gets updated?
Those are difficult problems indeed. What you mentioned about checking the secret maybe via the resourceID in a go-rountine could be enough in most cases. Alternatively, our tiller-proxy details could contain the sha or resourceID of the provided CustomCA, so we will compare it with the one cached. This will allow us to pre-load the data on boot but also re-fetch it if stale,

What I am worried here is about adding a synchronous operation to an already costly op (deploy or upgrade). Do we know how long it takes to pull the secret?

I am going to remove my "required changes" for this PR because it seems that we are just building on top of an already existing pattern and we are affected only when auth is set, but I'd love to see a follow up improvement here.
   @migmartri
migmartri 14 hours ago  Member
#895
   @prydonius
prydonius 13 hours ago  Member
Alternatively, our tiller-proxy details could contain the sha or resourceID of the provided CustomCA
Interesting idea, that would require the frontend to pull down the Secret to compute the checksum though, right? In that case, wouldn't it be okay to simply the pass the header and custom CA directly in the details? Assuming we talk to tiller-proxy over TLS, it's no less secure than us already passing the Kubernetes Bearer token in every request.
   @migmartri
migmartri 13 hours ago  Member
I am not completely familiar with how things work now but yes, somehow the sha would need to be generated and potentially stored in the appRepository CRD as annotation or field.

We could make the appRepository controller be in charge of populating that value too?

**Comment by andresmgot** [comment]
I think we are trying to solve a problem that doesn't exists. K8s already has a cache mechanism for storing results in the API server: https://github.com/kubernetes/kubernetes/issues/6678 so the impact of these requests would be minimal.

Note that we are fetching the chart for the registry when creating instances, that's way slower. We are also contacting the K8s API with all the requests to check that the user has permissions to perform a specific action. Since the requests are always the same, I believe those requests are already cached in the K8s API and the response time is minimal.

I don't think we should try to solve this ourselves unless we have specific data (metrics) that this is indeed a problem.

**Comment by prydonius** [comment]
Good point about the auth check, I agree that we should figure out what the actual bottleneck of this is. In general, it may be worth looking at the whole install flow to see if there's anyway we can optimise for what we're requesting, but I think this is pretty low priority.

**Comment by absoludity** [comment]
We've removed tiller-proxy, but the same argument would remain for kubeops.

So to get data, I ran an api server with `-v=3` to see requests, created a specific namespace and then grepped for the requests to the api server for that namespace while deploying apache:

```
$ k -n kube-system logs kube-apiserver-kubeapps-control-plane -f | grep test-foo-namespace | tee /tmp/895-test.log
I1001 02:33:42.178556       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-foo-namespace/resourcequotas" latency=867.859µs resp=200 UserAgent="kube-apiserver/v1.18.2 (linux/amd64) kubernetes/52c56ce" srcIP="[::1]:33046": 
I1001 02:33:42.179541       1 httplog.go:90] verb="POST" URI="/api/v1/namespaces/test-foo-namespace/serviceaccounts" latency=2.698168ms resp=201 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/system:serviceaccount:kube-system:service-account-controller" srcIP="172.18.0.2:60804": 
I1001 02:33:42.181588       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-foo-namespace/serviceaccounts/default" latency=1.629351ms resp=200 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:33:42.185067       1 httplog.go:90] verb="POST" URI="/api/v1/namespaces/test-foo-namespace/secrets" latency=1.336948ms resp=201 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:33:42.187114       1 httplog.go:90] verb="PUT" URI="/api/v1/namespaces/test-foo-namespace/serviceaccounts/default" latency=1.663663ms resp=200 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:33:42.246117       1 httplog.go:90] verb="GET" URI="/apis/operators.coreos.com/v1alpha1/namespaces/test-foo-namespace/clusterserviceversions" latency=123.525µs resp=404 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:59540": 
I1001 02:33:42.247485       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-foo-namespace" latency=1.564705ms resp=200 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:59528": 
I1001 02:33:42.250664       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-foo-namespace/secrets?labelSelector=owner%3Dhelm" latency=1.034872ms resp=200 UserAgent="kubeops/v0.0.0 (linux/amd64) kubernetes/$Format" srcIP="10.244.0.14:54758": 
I1001 02:34:02.222040       1 httplog.go:90] verb="GET" URI="/apis/operators.coreos.com/v1alpha1/namespaces/test-foo-namespace/clusterserviceversions" latency=528.817µs resp=404 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:60328": 
...(clipped remaining which were watching for the deployment)
```

If instead I first add an auth header for the repo and repeat with a different namespace, I see two extra requests for secrets:

```
I1001 02:44:23.757068       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace/resourcequotas" latency=12.266394ms resp=200 UserAgent="kube-apiserver/v1.18.2 (linux/amd64) kubernetes/52c56ce" srcIP="[::1]:33046": 
I1001 02:44:23.757984       1 httplog.go:90] verb="POST" URI="/api/v1/namespaces/test-bar-namespace/serviceaccounts" latency=14.123227ms resp=201 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/system:serviceaccount:kube-system:service-account-controller" srcIP="172.18.0.2:60804": 
I1001 02:44:23.767326       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace/serviceaccounts/default" latency=9.025264ms resp=200 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:44:23.770587       1 httplog.go:90] verb="POST" URI="/api/v1/namespaces/test-bar-namespace/secrets" latency=1.277009ms resp=201 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:44:23.773183       1 httplog.go:90] verb="PUT" URI="/api/v1/namespaces/test-bar-namespace/serviceaccounts/default" latency=2.189003ms resp=200 UserAgent="kube-controller-manager/v1.18.2 (linux/amd64) kubernetes/52c56ce/tokens-controller" srcIP="172.18.0.2:60612": 
I1001 02:44:23.816462       1 httplog.go:90] verb="GET" URI="/apis/operators.coreos.com/v1alpha1/namespaces/test-bar-namespace/clusterserviceversions" latency=70.326µs resp=404 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:44930": 
I1001 02:44:23.816913       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace" latency=906.566µs resp=200 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:44908": 
I1001 02:44:23.817254       1 httplog.go:90] verb="GET" URI="/apis/kubeapps.com/v1alpha1/namespaces/test-bar-namespace/apprepositories" latency=914.484µs resp=200 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:44924": 
I1001 02:44:23.818356       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace/secrets" latency=1.870293ms resp=200 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:44932": 
I1001 02:44:23.820206       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace/secrets?labelSelector=owner%3Dhelm" latency=966.265µs resp=200 UserAgent="kubeops/v0.0.0 (linux/amd64) kubernetes/$Format" srcIP="10.244.0.14:39892": 
I1001 02:44:23.874162       1 httplog.go:90] verb="GET" URI="/api/v1/namespaces/test-bar-namespace/secrets" latency=790.765µs resp=200 UserAgent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0" srcIP="10.244.0.12:44936": 
I
```

So I'm going to close as a non-issue for now. Re-open if you disagree :)
